src/bootloaders/shim-systemd.c

```diff
@@ -0,0 +1,313 @@
+        SHIM_SRC_DIR                                                                               \
+        "/"                                                                                        \
+        "fb" EFI_SUFFIX
+#define SYSTEMD_SRC_DIR "usr/lib/systemd/boot/efi"
+#define SYSTEMD_SRC                                                                                \
+        SYSTEMD_SRC_DIR                                                                            \
+        "/"                                                                                        \
+        "systemd-boot" EFI_SUFFIX
+#define DST_DIR "/" KERNEL_NAMESPACE
+#define SHIM_DST                                                                                   \
+        DST_DIR                                                                                    \
+        "/"                                                                                        \
+        "bootloader" EFI_SUFFIX
+#define SYSTEMD_DST                                                                                \
+        DST_DIR                                                                                    \
+        "/"                                                                                        \
+        "loader" EFI_SUFFIX
+#define KERNEL_DST_DIR DST_DIR "/kernel"
+#define SYSTEMD_CONFIG_DIR "/loader"
+#define SYSTEMD_CONFIG SYSTEMD_CONFIG_DIR "/loader.conf"
+#define SYSTEMD_ENTRIES SYSTEMD_CONFIG_DIR "/entries"
+
+static char *shim_src;
+static char *shim_dst_host; /* as accessible by the CMB for file ops. */
+static char *shim_dst_esp;  /* absolute location of shim on the ESP. */
+static char *systemd_src;
+static char *systemd_dst_host;
+
+static char *shim_systemd_get_kernel_dst(const BootManager *manager)
+{
+        (void)manager;
+        return strdup(KERNEL_DST_DIR);
+}
+
+static bool shim_systemd_install_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        return sd_class_install_kernel_impl(manager, kernel, shim_systemd_get_kernel_dst, NULL);
+}
+
+static bool shim_systemd_remove_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        return sd_class_remove_kernel(manager, kernel);
+}
+
+static bool shim_systemd_set_default_kernel(const BootManager *manager, const Kernel *kernel)
+{
+        /* this writes systemd config. systemd has the configuration paths
+         * hardcoded, hence whatever sd_class is doing is OK. */
+        return sd_class_set_default_kernel_impl(manager, kernel, NULL);
+}
+
+static bool exists_identical(const char *path, const char *spath)
+{
+        if (!nc_file_exists(path))
+                return false;
+        if (spath && !cbm_files_match(path, spath))
+                return false;
+        return true;
+}
+
+static bool shim_systemd_needs_install(const BootManager *manager)
+{
+        (void)manager;
```

Comment:
we have `__cbm_unused__` to mark a parameter as unused, no need for the `(void)` cast (which doesn't always work with older compilers)